File: zephyr_one/mobile/ios/Tests/ZephyrCoreTests/FakeKeyValueStore.swift
import Foundation
import XCTest
@testable import ZephyrCore

/// In-memory ``KeyValueStore``.
///
/// Models the two properties of the real thing the callers depend on: reads see
/// writes immediately, and a batch is applied as one unit. Values are held as `Any`
/// because the store is heterogeneous by design -- a share row is two strings plus a
/// boolean, a bookmark row is data plus a boolean -- and a typed dictionary per kind
/// would let a test pass while the production code read the wrong one.
final class FakeKeyValueStore: KeyValueStore {

    private var values: [String: Any]

    /// Number of batches applied. Proves a multi-key row is written once rather than
    /// key by key, where a termination mid-row would leave a partial row behind.
    private(set) var batches = 0

    init(values: [String: Any] = [:]) {
        self.values = values
    }

    func string(_ key: String) -> String? { values[key] as? String }

    func data(_ key: String) -> Data? { values[key] as? Data }

    func boolean(_ key: String, default defaultValue: Bool) -> Bool {
        values[key] as? Bool ?? defaultValue
    }

    func stringArray(_ key: String) -> [String] { values[key] as? [String] ?? [] }

    func keys() -> [String] { Array(values.keys) }

    func edit(_ block: (KeyValueWriter) -> Void) {
        batches += 1
        /* Staged and merged, so a batch that returns early leaves nothing behind.
         * UserDefaults behaves the same way from the caller's point of view. */
        let writer = Writer()
        block(writer)
        for key in writer.removed { values.removeValue(forKey: key) }
        for (key, value) in writer.staged { values[key] = value }
    }

    /// Simulates a relaunch: the same bytes, a new object graph on top of them.
    func surviveRestart() -> FakeKeyValueStore { FakeKeyValueStore(values: values) }

    /// Simulates external truncation, e.g. a partially cleared defaults domain.
    func drop(_ key: String) { values.removeValue(forKey: key) }

    private final class Writer: KeyValueWriter {
        var staged: [String: Any] = [:]
        var removed: Set<String> = []

        func setString(_ key: String, _ value: String) {
            staged[key] = value
            removed.remove(key)
        }

        func setData(_ key: String, _ value: Data) {
            staged[key] = value
            removed.remove(key)
        }

        func setBoolean(_ key: String, _ value: Bool) {
            staged[key] = value
            removed.remove(key)
        }

        func setStringArray(_ key: String, _ value: [String]) {
            staged[key] = value
            removed.remove(key)
        }

        func remove(_ key: String) {
            staged.removeValue(forKey: key)
            removed.insert(key)
        }
    }
}

File: zephyr_one/mobile/android/feature-file-sync/src/test/kotlin/one/zephyr/mobile/feature/filesync/FakeKeyValueStore.kt
package one.zephyr.mobile.feature.filesync

/**
 * In-memory [KeyValueStore].
 *
 * Models the two properties of the real thing that matter to the callers: reads see writes
 * immediately, and a batch is applied as one unit. Values are held as `Any` because the store is
 * heterogeneous by design -- a grant row is two strings plus a boolean plus a set -- and a typed map
 * per kind would let a test pass while the production code read the wrong one.
 */
class FakeKeyValueStore(
    private val values: MutableMap<String, Any> = LinkedHashMap(),
) : KeyValueStore {

    /** Makes the next durable batch fail without changing visible or restart state. */
    var failNextEdit: Boolean = false

    /** One-based batch number to fail; useful when an operation has a prepare and commit phase. */
    var failBatch: Int? = null

    /** Number of batches applied. Proves a multi-key row is written once, not key by key. */
    var batches = 0
        private set

    override fun string(key: String): String? = values[key] as? String

    override fun boolean(key: String, defaultValue: Boolean): Boolean =
        values[key] as? Boolean ?: defaultValue

    @Suppress("UNCHECKED_CAST")
    override fun stringSet(key: String): Set<String> = values[key] as? Set<String> ?: emptySet()

    override fun keys(): Set<String> = values.keys.toSet()

    override fun edit(block: KeyValueEditor.() -> Unit): Boolean {
        batches += 1
        /* Staged and then merged, so a batch that throws part-way leaves nothing behind. The real
         * editor behaves the same way: nothing is visible until apply(). */
        val staged = LinkedHashMap<String, Any>()
        val removed = mutableSetOf<String>()
        object : KeyValueEditor {
            override fun putString(key: String, value: String) {
                staged[key] = value
                removed -= key
            }

            override fun putBoolean(key: String, value: Boolean) {
                staged[key] = value
                removed -= key
            }

            override fun putStringSet(key: String, value: Set<String>) {
                staged[key] = value.toSet()
                removed -= key
            }

            override fun remove(key: String) {
                staged -= key
                removed += key
            }
        }.block()
        if (failNextEdit || failBatch == batches) {
            failNextEdit = false
            failBatch = null
            return false
        }
        for (key in removed) values.remove(key)
        values.putAll(staged)
        return true
    }

    /** Simulates a relaunch: the same bytes, a new object graph on top of them. */
    fun surviveRestart(): FakeKeyValueStore = FakeKeyValueStore(LinkedHashMap(values))

    /** Simulates external truncation, e.g. a partially cleared preferences file. */
    fun drop(key: String) {
        values.remove(key)
    }
}
